src/main/java/io/vertx/proton/impl/ProtonServerImpl.java

```diff
@@ -110,24 +110,60 @@ public void close(Handler<AsyncResult<Void>> handler) {
     return handler;
   }
 
+  public ProtonServer saslAuthenticator(ProtonSaslAuthenticator authenticator) {
+    if (authenticator == null) {
+      // restore the default authenticator
+      this.authenticator = new ProtonSaslServerAuthenticatorImpl();
+    } else {
+      this.authenticator = authenticator;
+    }
+    return this;
+  }
+
   public ProtonServerImpl connectHandler(Handler<ProtonConnection> handler) {
     this.handler = handler;
-    server.connectHandler(new Handler<NetSocket>() {
-      @Override
-      public void handle(NetSocket netSocket) {
-        String hostname = null;
-        try {
-          hostname = InetAddress.getLocalHost().getHostName();
-        } catch (UnknownHostException e) {
+    server.connectHandler(netSocket -> {
+      String hostname = null;
+      try {
+        hostname = InetAddress.getLocalHost().getHostName();
+      } catch (UnknownHostException e) {
+      }
+
+      final ProtonConnectionImpl connection = new ProtonConnectionImpl(vertx, hostname);
+      if (advertiseAnonymousRelayCapability) {
+        connection.setOfferedCapabilities(new Symbol[] { ProtonConnectionImpl.ANONYMOUS_RELAY });
+      }
+
+      connection.bindServer(netSocket, new ProtonSaslAuthenticator () {
+
+        @Override
+        public void init(Transport transport) {
+          authenticator.init(transport);
         }
 
-        ProtonConnectionImpl connection = new ProtonConnectionImpl(vertx, hostname);
-        if (advertiseAnonymousRelayCapability) {
-          connection.setOfferedCapabilities(new Symbol[] { ProtonConnectionImpl.ANONYMOUS_RELAY });
+        @Override
+        public boolean process() {
+          boolean result = authenticator.process();
+          if (result) {
+            // the authenticator completed
+            if (succeeded()) {
+              handler.handle(connection);
+            } else {
+              // auth failed, disconnect client
+              connection.disconnect();
```

Comment:
The connection needs flushed before disconnection, so that the failure details are actually sent to the client.